src/graph.py

```diff
@@ -0,0 +1,66 @@
+import itertools
+
+
+class Graph():
+    def __init__(self):
+        # Create empty graph
+        self._container = {}
+
+    def nodes(self):
+        """Return a list of all nodes."""
+        return [node for node in self._container]
+
+    def edges(self):
+        """Retun a list of all edges."""
+        edges = []
+        for node in self._container:
+            for value in self._container[node]:
+                edges.append((node, value))
+        return edges
+
+    def add_node(self, n):
+        """Add node to graph."""
+        self._container.setdefault(n, set())
+
+    def add_edge(self, n1, n2):
+        """Add edge connecting node n1 to node n2.
+
+        Creates nodes n1, n2 if necessary.
+        """
+        self.add_node(n2)
+        self._container.setdefault(n1, set()).add(n2)
+
+    def del_node(self, n):
+        """Delete node."""
+        if self.has_node(n):
```

Comment:
g.del_node() doesn't necessarily need a conditional. If you want to add some exception-handling, then you can do a try...except... bit. Note however that deleting an item from a dictionary will automatically throw a KeyError if the key isn't there.